zookeeper-server/src/main/java/org/apache/zookeeper/Login.java

```diff
@@ -95,10 +95,159 @@
+                    while (retry >= 0) {
+                        try {
+                            LOG.debug("running ticket cache refresh command: {} {}", cmd, kinitArgs);
+                            Shell.execCommand(cmd, kinitArgs);
+                            break;
+                        } catch (Exception e) {
+                            if (retry > 0) {
+                                --retry;
+                                // sleep for 10 seconds
+                                try {
+                                    Thread.sleep(10 * 1000);
+                                } catch (InterruptedException ie) {
+                                    LOG.error("Interrupted while renewing TGT, exiting Login thread");
+                                    return;
+                                }
+                            } else {
+                                LOG.warn(
+                                    "Could not renew TGT due to problem running shell command: '{} {}'."
+                                        + " Exiting refresh thread.",
+                                    cmd,
+                                    kinitArgs,
+                                    e);
+                                return;
+                            }
+                        }
+                    }
+                }
+                try {
+                    int retry = 1;
+                    while (retry >= 0) {
+                        try {
+                            reLogin();
+                            break;
+                        } catch (LoginException le) {
+                            if (retry > 0) {
+                                --retry;
+                                // sleep for 10 seconds.
+                                try {
+                                    Thread.sleep(10 * 1000);
+                                } catch (InterruptedException e) {
+                                    LOG.error("Interrupted during login retry after LoginException:", le);
+                                    throw le;
+                                }
+                            } else {
+                                LOG.error("Could not refresh TGT for principal: {}.", principal, le);
+                            }
+                        }
+                    }
+                } catch (LoginException le) {
+                    LOG.error("Failed to refresh TGT: refresh thread exiting now.", le);
+                    break;
+                }
+            }
+        }
+    }
+
+    public void loginAndStartThreadIfRequired() throws LoginException {
         login = login(loginContextName);
-        this.loginContextName = loginContextName;
         subject = login.getSubject();
         isKrbTicket = !subject.getPrivateCredentials(KerberosTicket.class).isEmpty();
         AppConfigurationEntry[] entries = Configuration.getConfiguration().getAppConfigurationEntry(loginContextName);
```

Comment:
You moved this logic from the constructor to the new method `loginAndStartThreadIfRequired()` which could be beneficial if the constuction of the Login object and the method call would be separated, but I don't see that would be the case.

Login class creation always goes together with startIfNeeded(), so it doesn't really matter what happens in which. I'd like to see what exactly the resource which is not cleared properly. The thread object?

Reply:
Yes, the refresh thread of Login object is not cleaned up leaving many login threads.
The Jira's description is missing one key stack trace

```
17971 Threads with the following stack trace

3911 - TIMED_WAITING - Thread-822 | ORG: null RAC: None
java.base@11.0.9/java.lang.Thread.sleep(Native Method)
org.apache.zookeeper.Login$1.run(Login.java:197)
java.base@11.0.9/java.lang.Thread.run(Thread.java:834)
```

which when combined with another stack trace coming from ZookeeperSaslClient

```
java.lang.Thread.State: RUNNABLE at java.net.PlainSocketImpl.socketConnect(java.base@11.0.4.0.101/Native Method) at java.net.AbstractPlainSocketImpl.doConnect(java.base@11.0.4.0.101/AbstractPlainSocketImpl.java:399) - locked <0x00000015004fde20> (a java.net.SocksSocketImpl) at java.net.AbstractPlainSocketImpl.connectToAddress(java.base@11.0.4.0.101/AbstractPlainSocketImpl.java:242) at java.net.AbstractPlainSocketImpl.connect(java.base@11.0.4.0.101/AbstractPlainSocketImpl.java:224) at java.net.SocksSocketImpl.connect(java.base@11.0.4.0.101/SocksSocketImpl.java:403) at java.net.Socket.connect(java.base@11.0.4.0.101/Socket.java:609) at sun.security.krb5.internal.TCPClient.<init>(java.security.jgss@11.0.4.0.101/NetClient.java:62) at sun.security.krb5.internal.NetClient.getInstance(java.security.jgss@11.0.4.0.101/NetClient.java:42) at sun.security.krb5.KdcComm$KdcCommunication.run(java.security.jgss@11.0.4.0.101/KdcComm.java:401) at sun.security.krb5.KdcComm$KdcCommunication.run(java.security.jgss@11.0.4.0.101/KdcComm.java:364) at java.security.AccessController.doPrivileged(java.base@11.0.4.0.101/Native Method) at sun.security.krb5.KdcComm.send(java.security.jgss@11.0.4.0.101/KdcComm.java:348) at sun.security.krb5.KdcComm.sendIfPossible(java.security.jgss@11.0.4.0.101/KdcComm.java:253) at sun.security.krb5.KdcComm.send(java.security.jgss@11.0.4.0.101/KdcComm.java:234) at sun.security.krb5.KdcComm.send(java.security.jgss@11.0.4.0.101/KdcComm.java:200) at sun.security.krb5.KrbAsReqBuilder.send(java.security.jgss@11.0.4.0.101/KrbAsReqBuilder.java:326) at sun.security.krb5.KrbAsReqBuilder.action(java.security.jgss@11.0.4.0.101/KrbAsReqBuilder.java:371) at com.sun.security.auth.module.Krb5LoginModule.attemptAuthentication(jdk.security.auth@11.0.4.0.101/Krb5LoginModule.java:754) at com.sun.security.auth.module.Krb5LoginModule.login(jdk.security.auth@11.0.4.0.101/Krb5LoginModule.java:592) at javax.security.auth.login.LoginContext.invoke(java.base@11.0.4.0.101/LoginContext.java:726) at javax.security.auth.login.LoginContext$4.run(java.base@11.0.4.0.101/LoginContext.java:665) at javax.security.auth.login.LoginContext$4.run(java.base@11.0.4.0.101/LoginContext.java:663) at java.security.AccessController.doPrivileged(java.base@11.0.4.0.101/Native Method) at javax.security.auth.login.LoginContext.invokePriv(java.base@11.0.4.0.101/LoginContext.javaa:663) at javax.security.auth.login.LoginContext.login(java.base@11.0.4.0.101/LoginContext.java:574) at org.apache.zookeeper.Login.login(Login.java:304) - locked <0x000000151c477148> (a org.apache.zookeeper.Login) at org.apache.zookeeper.Login.<init>(Login.java:106) at org.apache.zookeeper.client.ZooKeeperSaslClient.createSaslClient(ZooKeeperSaslClient.java:249) - locked <0x000000151c476f68> (a org.apache.zookeeper.client.ZooKeeperSaslClient) at org.apache.zookeeper.client.ZooKeeperSaslClient.<init>(ZooKeeperSaslClient.java:141) at org.apache.zookeeper.ClientCnxn$SendThread.startConnect(ClientCnxn.java:972) at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1031)
```

Here is one possibility of Login threads may not cleared by shutdown()

1. Zookeeper initialization creates & starts ClientCnxn -> spins up SendThread and starts it
2. ClientCnxn.SendThread.run() calls ZookeeperSaslClient.<init> -> createSaslClient()
3. Login.<init> -> Login.login() -> loginContext.login()
4. Let's say thread1 executing a login (loginContext.login()) takes few seconds.
5. Meantime, let's say some one called Zookeeper.close() -> ClientCnxn.close() -> ClientCnxn.disconnect()
6. At this point, since ZooKeeperSaslClient's constructor is not yet finished, it's zooKeeperSaslClient.shutdown() is never called.
7. Once 3 get successful, a login thread refresh is created which would become a zombie.

So, the idea is 2 parts
1. Move the actual "login" which can take time in to separate methods and make ZooKeeperSaslClient and Login constructors lightweight so that we have a reference to these at all times.
2. Make sure either Login thread is interrupted (when shutdown called after the thread was started) OR not started (when shutdown called before the thread was started)

Reply:
@anmolnar , could you please review the updated files with above context?

Reply:
Doing another round of review, sorry for the delay.